### PR TITLE
Fixes zombie movements #3

### PR DESCRIPTION
**Pull request by fredy001**

### CODE REVIEW

File: src/creeps/Zombie.cpp [hunk 1]
```diff
@@ -57,38 +60,158 @@ bool Zombie::isMoving() const {
...
+        objId = 3;
+    } else if (ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeBarricade,this),this)) {
+        objId = 4;
+    }
+    
+    /*return (ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeMarine, this), this)
             || ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeObj, this), this)
-            || ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeTurret, this), this));
+            || ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeBarricade,this),this)
+            || ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeTurret, this), this));*/
```
Comment (RobertArendac) [inline src/creeps/Zombie.cpp]: Instead of commenting out unused code, you could just get rid of it entirely

File: src/creeps/Zombie.cpp [hunk 2]
```diff
@@ -57,38 +60,158 @@ bool Zombie::isMoving() const {
...
+ * overriden move method, preventing zombies from blocking
+ * Fred Yang,  Robert Arendac
+ * March 15
+*/
+void Zombie::move(float moveX, float moveY, CollisionHandler& ch){
+    ZombieDirection newDir = ZombieDirection::DIR_INVALID;
+    ZombieDirection nextDir = ZombieDirection::DIR_INVALID;
+
+    float oldX = getX();
+    float oldY = getY();
```
Comment (RobertArendac) [inline src/creeps/Zombie.cpp]: Make things const when you can

File: src/creeps/Zombie.cpp [hunk 3]
```diff
@@ -57,38 +60,158 @@ bool Zombie::isMoving() const {
...
+    setX(getX() + moveX);
+
+    if (ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeMarine,this),this)
+            || ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeZombie,this),this)
+            || ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeWall,this),this)
+            || ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeBarricade,this),this)
+            || ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeTurret,this),this)
+            || ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeObj,this),this)) {
+        setX(getX() - moveX);
+    }
```
Comment (RobertArendac) [inline src/creeps/Zombie.cpp]: You made a method to check if the zombie hit something, why not use that here?

File: src/creeps/Zombie.cpp [hunk 4]
```diff
@@ -57,38 +60,158 @@ bool Zombie::isMoving() const {
...
+    setY(getY()+moveY);
+
+    if (ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeMarine,this),this)
+            || ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeZombie,this),this)
+            || ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeWall,this),this)
+            || ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeBarricade,this),this)
+            || ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeTurret,this),this)
+            || ch.detectMovementCollision(ch.getQuadTreeEntities(ch.quadtreeObj,this),this)) {
+        setY(getY() - moveY);
+    }
```
Comment (RobertArendac) [inline src/creeps/Zombie.cpp]: Same as above, why not use your method?

File: src/creeps/Zombie.cpp [hunk 5]
```diff
@@ -57,38 +60,158 @@ bool Zombie::isMoving() const {
...
+
+    float curX = getX();
+    float curY = getY();
+    float dist = sqrt((curX - oldX)*(curX - oldX) + (curY - oldY)*(curY - oldY));
```
Comment (RobertArendac) [inline src/creeps/Zombie.cpp]: Make const

File: src/creeps/Zombie.cpp [hunk 6]
```diff
@@ -153,72 +276,87 @@ void Zombie::generateMove() {
 /**
  * A* algo generates a string of direction digits.
  * Fred Yang
+ * March 15
+ */
+string Zombie::generatePath(const Point& start) {
+    return generatePath(start, Point(MAP_WIDTH/2, MAP_HEIGHT/2));
+}
```
Comment (RobertArendac) [inline src/creeps/Zombie.cpp]: Is this function necessary if it's just going to call essentially the same function?

File: src/creeps/Zombie.cpp [hunk 7]
```diff
@@ -153,72 +276,87 @@ void Zombie::generateMove() {
...
     // path to be generated
     string path;
 
+    // current node & child node
+    static Node curNode;
+    static Node childNode;
+
     // priority queue
-    static array<priority_queue<Node>, 2> pq;
+    static priority_queue<Node> pq[2];
```
Comment (RobertArendac) [inline src/creeps/Zombie.cpp]: Why change this from what was there before?

File: src/creeps/Zombie.cpp [hunk 8]
```diff
@@ -153,72 +276,87 @@ void Zombie::generateMove() {
...
+    memset(dirMap, 0, sizeof(int) * ROWS * COLS);
 
-    const int xNodeStart = static_cast<int>(xStart / TILE_SIZE);
-    const int yNodeStart = static_cast<int>(yStart / TILE_SIZE);
-    const int xNodeDest = static_cast<int>(xDest / TILE_SIZE);
-    const int yNodeDest = static_cast<int>(yDest / TILE_SIZE);
+    int xNodeStart = static_cast<int> (start.second + TILE_OFFSET) / TILE_SIZE;
+    int yNodeStart = static_cast<int> (start.first + TILE_OFFSET) / TILE_SIZE;
+    int xNodeDest = static_cast<int> (dest.second + TILE_OFFSET) / TILE_SIZE - 1;
+    int yNodeDest = static_cast<int> (dest.first + TILE_OFFSET) / TILE_SIZE - 1;
```
Comment (RobertArendac) [inline src/creeps/Zombie.cpp]: Why remove the const for each variable?

File: src/creeps/Zombie.cpp [hunk 9]
```diff
@@ -153,72 +276,87 @@ void Zombie::generateMove() {
...
+    // row & column index
+    int curRow, curCol;
+    int newRow, newCol;
 
     // path to be generated
     string path;
 
+    // current node & child node
+    static Node curNode;
+    static Node childNode;
```
Comment (RobertArendac) [inline src/creeps/Zombie.cpp]: Do these need to be static?

File: src/creeps/Zombie.cpp [hunk 10]
```diff
@@ -153,72 +276,87 @@ void Zombie::generateMove() {
...
-    const int yNodeDest = static_cast<int>(yDest / TILE_SIZE);
+    int xNodeStart = static_cast<int> (start.second + TILE_OFFSET) / TILE_SIZE;
+    int yNodeStart = static_cast<int> (start.first + TILE_OFFSET) / TILE_SIZE;
+    int xNodeDest = static_cast<int> (dest.second + TILE_OFFSET) / TILE_SIZE - 1;
+    int yNodeDest = static_cast<int> (dest.first + TILE_OFFSET) / TILE_SIZE - 1;
 
-    // create the start node and push the start node into open list
-    Node curNode(xNodeStart, yNodeStart);
+    // create the start node and push into open list
+    curNode = Node(xNodeStart, yNodeStart, 0, 0);
```
Comment (RobertArendac) [inline src/creeps/Zombie.cpp]: Why use copying when you could've used the constructor when declaring it?

File: src/creeps/Zombie.cpp [hunk 11]
```diff
@@ -153,72 +276,87 @@ void Zombie::generateMove() {
...
     curNode.updatePriority(xNodeDest, yNodeDest);
     pq[index].push(curNode);
 
     // A* path finding
     while (!pq[index].empty()) {
         // get the current node with the highest priority from open list
-        curNode = pq[index].top();
+        curNode = Node(pq[index].top().getXPos(), pq[index].top().getYPos(),
+                       pq[index].top().getLevel(), pq[index].top().getPriority());
```
Comment (RobertArendac) [inline src/creeps/Zombie.cpp]: Why can't this be `curNode = pq[index].top()`?

File: src/creeps/Zombie.cpp [hunk 12]
```diff
@@ -153,72 +276,87 @@ void Zombie::generateMove() {
...
             path = "";
-            while (!(x == xNodeStart && y == yNodeStart)) {
-                j = dirMap[x][y];
-                path = static_cast<char>('0' + (j + DIR_CAP / 2) % DIR_CAP) + path;
-                x += MX[j];
-                y += MY[j];
+            while (!(curRow == xNodeStart && curCol == yNodeStart)) {
+                j = dirMap[curRow][curCol];
+                c = '0' + (j + DIR_CAP/2)%DIR_CAP;
+                path = c + path;
```
Comment (RobertArendac) [inline src/creeps/Zombie.cpp]: Why do you need the `c` variable if you only use it here?

File: src/creeps/Zombie.cpp [hunk 13]
```diff
@@ -153,72 +276,87 @@ void Zombie::generateMove() {
...
+                curRow += MY[j];
+                curCol += MX[j];
             }
 
             // empty the leftover nodes
-            pq[index] = priority_queue<Node>();
+            while (!pq[index].empty()) {
+                 pq[index].pop();
+            }
```
Comment (RobertArendac) [inline src/creeps/Zombie.cpp]: Why change this?  They both do the same thing but `pq[index] = priority_queue<Node>()` was a quicker and cleaner solution.

File: src/creeps/Zombie.cpp [hunk 14]
```diff
@@ -227,49 +365,51 @@ string Zombie::generatePath(const float xStart, const float yStart,
...
 
             // not evaluated & not outside (bound checking)
-            if (!(xdx < 0 || xdx > COL - 1 || ydy < 0 || ydy > ROW - 1
-                    || gameMap[xdx][ydy] == 1 || closedNodes[xdx][ydy] == 1)) {
+            if (!(newRow < 0 || newRow > COLS - 1 || newCol < 0 || newCol > ROWS - 1
+                || gameMap[newRow][newCol] >= 1 || closedNodes[newRow][newCol] == 1)) {
 
                 // generate a child node
-                Node childNode(xdx, ydy, curNode.getLevel(), curNode.getPriority());
+                childNode = Node(newRow, newCol, curNode.getLevel(), curNode.getPriority());
```
Comment (RobertArendac) [inline src/creeps/Zombie.cpp]: Why change this?

File: src/creeps/Zombie.cpp [hunk 15]
```diff
@@ -227,49 +365,51 @@ string Zombie::generatePath(const float xStart, const float yStart,
...
 
-                    // switch to pq with smaller size
-                    if (pq[index].size() > pq[1 - index].size()) {
+                    if (pq[index].size() > pq[1-index].size()) {
                         index = 1 - index;
                     }
 
                     while (!pq[index].empty()) {
-                        pq[1 - index].push(pq[index].top());
+                        pq[1-index].push(pq[index].top());
```
Comment (RobertArendac) [inline src/creeps/Zombie.cpp]: Why take out the spacing?

**Review by RobertArendac — CHANGES_REQUESTED** [review]
Some styling changes required.

**Comment by fredy001** [comment]
made changes per feedback

File: src/creeps/Zombie.cpp [hunk 16]
```diff
@@ -57,38 +60,148 @@ bool Zombie::isMoving() const {
...
         if (frame > 0) {
             --frame;
         }
 
-        // Changed to attack state once attack code is ready
-        setState(ZombieState::ZOMBIE_IDLE);
-
+        if (collisionObjId > 5) { // base
+            setState(ZombieState::ZOMBIE_IDLE);
+        } else if (collisionObjId > 2) { // marine, turret, or barricade
```
Comment (RobertArendac) [inline src/creeps/Zombie.cpp]: Base is something the zombie is going to attack as well, you can just change this thing to:
```
if (collisionObjId > 2)
    setState(ZombieState::ZOMBIE_ATTACK);

**Review by RobertArendac — CHANGES_REQUESTED** [review]
Just one small change I'd like you to make

**Pull request merged**